security-oauth2/src/main/java/io/micronaut/security/oauth2/endpoint/token/request/password/ReactiveOauthPasswordAuthenticationProvider.java

```diff
@@ -0,0 +1,95 @@
+import io.micronaut.security.authentication.AuthenticationResponse;
+import io.micronaut.security.authentication.provider.ReactiveAuthenticationProvider;
+import io.micronaut.security.oauth2.configuration.OauthClientConfiguration;
+import io.micronaut.security.oauth2.configuration.endpoints.SecureEndpointConfiguration;
+import io.micronaut.security.oauth2.endpoint.AuthenticationMethod;
+import io.micronaut.security.oauth2.endpoint.DefaultSecureEndpoint;
+import io.micronaut.security.oauth2.endpoint.SecureEndpoint;
+import io.micronaut.security.oauth2.endpoint.token.request.TokenEndpointClient;
+import io.micronaut.security.oauth2.endpoint.token.request.context.OauthPasswordTokenRequestContext;
+import io.micronaut.security.oauth2.endpoint.token.response.OauthAuthenticationMapper;
+import java.util.Collections;
+import java.util.List;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+
+/**
+ * An {@link ReactiveAuthenticationProvider} that delegates to an OAuth 2.0 provider using the
+ * password grant flow.
+ *
+ * @author Sergio del Amo
+ * @since 4.8.0
+ * @param <T> Request Context Type
+ * @param <I> Authentication Request Identity Type
+ * @param <S> Authentication Request Secret Type
+ */
+public class ReactiveOauthPasswordAuthenticationProvider<T, I, S> implements ReactiveAuthenticationProvider<T, I, S> {
+
+    private final TokenEndpointClient tokenEndpointClient;
+    private final SecureEndpoint secureEndpoint;
+    private final OauthClientConfiguration clientConfiguration;
+    private final OauthAuthenticationMapper authenticationMapper;
+
+    /**
+     * @param tokenEndpointClient The token endpoint client
+     * @param clientConfiguration The client configuration
+     * @param authenticationMapper  The user details mapper
+     */
+    public ReactiveOauthPasswordAuthenticationProvider(TokenEndpointClient tokenEndpointClient,
+                                                       OauthClientConfiguration clientConfiguration,
+                                                       OauthAuthenticationMapper authenticationMapper) {
+        this.tokenEndpointClient = tokenEndpointClient;
+        this.clientConfiguration = clientConfiguration;
+        this.authenticationMapper = authenticationMapper;
+        this.secureEndpoint = getTokenEndpoint(clientConfiguration);
+    }
+
+    @Override
+    public Publisher<AuthenticationResponse> authenticate(T requestContext, AuthenticationRequest<I, S> authenticationRequest) {
+
+        OauthPasswordTokenRequestContext context = new OauthPasswordTokenRequestContext(authenticationRequest, secureEndpoint, clientConfiguration);
+
+        return Flux.from(
+                        tokenEndpointClient.sendRequest(context))
+                .switchMap(response -> Flux.from(authenticationMapper.createAuthenticationResponse(response, null))
+                        .map(AuthenticationResponse.class::cast));
```

Comment:
Is there a particular motivation for using `switchMap` here?

I feel like it could be simplified to:
```
return Flux.from(tokenEndpointClient.sendRequest(context))
           .flatMap(response -> authenticationMapper.createAuthenticationResponse(response, null));
```

Reply:
> Is there a particular motivation for using `switchMap` here?

`ReactiveOauthPasswordAuthenticationProvider` is a port of  .https://github.com/micronaut-projects/micronaut-security/blob/4.8.x/security-oauth2/src/main/java/io/micronaut/security/oauth2/endpoint/token/request/password/OauthPasswordAuthenticationProvider.java which used switch mapped since its creation. 

https://stackoverflow.com/questions/28175702/what-is-the-difference-between-flatmap-and-switchmap-in-rxjava

Reply:
@jeremyg484 please, create a PR with this one as based with your recommended changes

Reply:
See #1694